File: backend/DevTools.Domain/Entities/Shortcut.cs
namespace DevTools.Domain.Entities;

public class Shortcut
{
    public Guid Id { get; private set; }
    public string Title { get; private set; }
    public string Keys { get; private set; }
    public string? Description { get; private set; }
    public string? Tags { get; private set; }
    public bool IsFavorite { get; private set; }
    public DateTimeOffset CreatedAt { get; private set; }
    public Guid CategoryId { get; private set; }
    public Guid UserId { get; private set; }

    private Shortcut()
    {
        Title = string.Empty;
        Keys = string.Empty;
    }

    public Shortcut(string title, string keys, string? description, string? tags, Guid categoryId, Guid userId)
    {
        Id = Guid.NewGuid();
        Title = title;
        Keys = keys;
        Description = description;
        Tags = tags;
        IsFavorite = false;
        CreatedAt = DateTimeOffset.UtcNow;
        CategoryId = categoryId;
        UserId = userId;
    }

    public void Update(string title, string keys, string? description, string? tags, Guid categoryId)
    {
        Title = title;
        Keys = keys;
        Description = description;
        Tags = tags;
        CategoryId = categoryId;
    }

    public void SetFavorite(bool isFavorite) => IsFavorite = isFavorite;
}

File: web/src/app/core/services/shortcut.ts
import { HttpClient } from '@angular/common/http';
import { Service, inject, signal } from '@angular/core';
import { firstValueFrom } from 'rxjs';

import { API_BASE_URL } from '../config/api-config';
import type { Shortcut, ShortcutRequest } from '../models/shortcut.model';

@Service()
export class ShortcutService {
  private readonly http = inject(HttpClient);

  private readonly shortcutsSignal = signal<Shortcut[]>([]);
  readonly shortcuts = this.shortcutsSignal.asReadonly();

  async load(): Promise<void> {
    const shortcuts = await firstValueFrom(this.http.get<Shortcut[]>(`${API_BASE_URL}/shortcuts`));
    this.shortcutsSignal.set(shortcuts);
  }

  async create(request: ShortcutRequest): Promise<Shortcut> {
    const shortcut = await firstValueFrom(
      this.http.post<Shortcut>(`${API_BASE_URL}/shortcuts`, request),
    );
    this.shortcutsSignal.update((list) => [shortcut, ...list]);
    return shortcut;
  }

  async update(id: string, request: ShortcutRequest): Promise<Shortcut> {
    const updated = await firstValueFrom(
      this.http.put<Shortcut>(`${API_BASE_URL}/shortcuts/${id}`, request),
    );
    this.shortcutsSignal.update((list) => list.map((s) => (s.id === id ? updated : s)));
    return updated;
  }

  async setFavorite(id: string, isFavorite: boolean): Promise<void> {
    const updated = await firstValueFrom(
      this.http.patch<Shortcut>(`${API_BASE_URL}/shortcuts/${id}/favorite`, { isFavorite }),
    );
    this.shortcutsSignal.update((list) => list.map((s) => (s.id === id ? updated : s)));
  }

  async delete(id: string): Promise<void> {
    await firstValueFrom(this.http.delete<void>(`${API_BASE_URL}/shortcuts/${id}`));
    this.shortcutsSignal.update((list) => list.filter((s) => s.id !== id));
  }

  /** Re-creates a just-deleted shortcut, for the "undo" snackbar action. */
  async restore(shortcut: Shortcut): Promise<void> {
    await this.create({
      title: shortcut.title,
      keys: shortcut.keys,
      description: shortcut.description,
      categoryId: shortcut.categoryId,
      tags: shortcut.tags,
    });
  }
}
